src/test/java/com/artipie/asto/s3/S3StorageTest.java

```diff
@@ -252,4 +259,35 @@ private S3Storage storage() {
             .build();
         return new S3Storage(client, this.bucket);
     }
+
+    /**
+     * Publisher that produces value only once for first subscription.
+     *
+     * @since 0.19
+     */
+    private static class OneOffPublisher implements Publisher<ByteBuffer> {
+
+        /**
+         * Data for subscriber.
+         */
+        private final ByteBuffer data;
+
+        /**
+         * Flag for completion.
+         */
+        private volatile boolean complete;
+
+        OneOffPublisher(final ByteBuffer data) {
+            this.data = data;
+        }
+
+        @Override
+        public void subscribe(final Subscriber<? super ByteBuffer> subscriber) {
+            if (!this.complete) {
```

Comment:
@olegmoz publisher may be subscribed from different threads, so I'd use `AtomicBoolean complete` and
```java
if (this.complete.compareAndSEt(false, true)) {
...
}
```

Reply:
@g4s8 you are right, fixed that